sklearn/preprocessing/tests/test_polynomial.py

```diff
@@ -490,6 +495,101 @@ def test_spline_transformer_n_features_out(
+        spline.fit(X_nan, sample_weight=sample_weight)
+
+
+@pytest.mark.skipif(
+    sp_version < parse_version("1.8.0"),
+    reason="The option `sparse_output` is available as of scipy 1.8.0",
+)
+@pytest.mark.parametrize(
+    "extrapolation", ["error", "constant", "linear", "continue", "periodic"]
+)
+@pytest.mark.parametrize("sparse_output", [False, True])
+def test_spline_transformer_handles_missing_values(extrapolation, sparse_output):
+    """Test that SplineTransformer handles missing values correctly.
+    We only test for knots="uniform", since for "quantile" the metrics are calculated
+    differently with nans present and a different result is thus expected.
+    """
+    X_nan = np.array([[1, 1], [2, 2], [3, 3], [np.nan, 5], [4, 4]])
+    X_nan_full_column = np.array([[np.nan, np.nan], [np.nan, 1]])
+    X = np.array([[1, 1], [2, 2], [3, 3], [4, 5], [4, 4]])
+
+    # check correct error message for handle_missing="error"
+    msg = "X contains missing values (np.nan) and SplineTransformer is configured with"
+    with pytest.raises(ValueError, match=re.escape(msg)):
+        spline = SplineTransformer(
+            degree=2,
+            n_knots=3,
+            handle_missing="error",
+            extrapolation=extrapolation,
+        )
+        spline.fit_transform(X_nan)
+
+    # check correct results for handle_missing="zeros"
+    spline = SplineTransformer(
+        degree=2,
+        n_knots=3,
+        handle_missing="zeros",
+        extrapolation=extrapolation,
+        sparse_output=sparse_output,
+    )
+
+    # check for generic invariants
+    X_nan_transform = spline.fit_transform(X_nan)
+    if sparse.issparse(X_nan_transform):
+        X_nan_transform = X_nan_transform.toarray()
+    assert (X_nan_transform >= 0).all()
+    assert (X_nan_transform <= 1).all()
+
+    # check `fit_transform` does the same as `fit` and `then` transform
+    X_nan_transform = spline.fit_transform(X_nan)
+    X_nan_fit_then_transformed = spline.fit(X_nan).transform(X_nan)
+    assert_allclose_dense_sparse(X_nan_transform, X_nan_fit_then_transformed)
+
+    # check that transform works as expected when the passed data has not the same
+    # shape as the training set array:
+    X_transformed_same_shape = spline.fit_transform(X_nan)[::2]
+    X_transformed_different_shapes = spline.fit(X_nan).transform(X_nan[::2])
+    assert_allclose_dense_sparse(
+        X_transformed_same_shape, X_transformed_different_shapes
+    )
+
+    # check that if X has a feature of all nans SplineTransformer works as usual
+    spline.transform(X_nan_full_column)
+
+    # check that the masked nan-values are 0s
+    nan_mask = _get_mask(X_nan, np.nan)
+    encoded_nan_mask = np.repeat(nan_mask, spline.bsplines_[0].c.shape[1], axis=1)
+    assert (X_nan_transform[encoded_nan_mask] == 0).all()
+
+    # check that nan values are always encoded as zeros, even in columns where
+    # no missing values were observed at training time.
```

Comment:
This comment does not seem to reflect what is being tested below: here `X_nan_full_column` has 2 columns, the fist one is all-nan and the second column has a mix of missing an non-missing values.

The comment seem to refer to the test case I proposed in https://github.com/scikit-learn/scikit-learn/pull/28043/files#r1592724271 which is different: check how nan values are encoded at test time in columns that have no missing value at training time (e.g. the second column of `X_nan`).

Reply:
I can see how this is confusing. I will try to re-formulate this in the new test below (`test_spline_transformer_handles_all_nans`), as discussed above. The attempt is to pull those two test groups apart. Is it clearer and consistent like this?

(The history of this here is that I have taken the test you have proposed and modified it.)